### PR TITLE
AEA-1292 Add other language support to the generator (protobuf only mode)

### PR DESCRIPTION
**Pull request by 5A11**
This PR adds other language support to the "protobuf" mode of the protocol generator. 

You can now do `aea generate protocol --l cpp <path_to_spec>` to generate protocol in protobuf only mode for c++ language. The language choices are: 

```
python, go, cpp, java, csharp, ruby, objc
```

For `go` an additional component needs to be installed for the `protoc` compiler. Instructions are in the protobuf website.

#2040 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/cli/generate.py [hunk 1]
```diff
@@ -45,44 +50,62 @@ def generate(
 
 @generate.command()
 @click.argument("protocol_specification_path", type=str, required=True)
+@click.option(
+    "--l",
+    "language",
+    type=str,
```
Comment (DavidMinarsch) [inline aea/cli/generate.py:56]:
```suggestion
    type=click.Choice(SUPPORTED_PROTOCOL_LANGUAGES)
```

where `SUPPORTED_PROTOCOL_LANGUAGES` is a list defined in constants.

File: aea/cli/generate.py [hunk 2]
```diff
@@ -45,44 +50,62 @@ def generate(
...
+    required=False,
+    default="python",
+    help="Specify the language in which to generate the protocol package.",
+)
 @pass_ctx
-def protocol(ctx: Context, protocol_specification_path: str) -> None:
+def protocol(ctx: Context, protocol_specification_path: str, language: str) -> None:
     """Generate a protocol based on a specification and add it to the configuration file and agent."""
-    _generate_item(ctx, PROTOCOL, protocol_specification_path)
```
Comment (DavidMinarsch) [inline aea/cli/generate.py:51]: Reintroduce that separation! It helps reuse of the function outside the cli.

File: aea/cli/generate.py [hunk 3]
```diff
@@ -91,23 +114,37 @@ def _generate_item(ctx: Context, item_type: str, specification_path: str) -> Non
...
+    agent_name = ctx.agent_config.agent_name
+    click.echo(
+        "Generating {} '{}' and adding it to the agent '{}'...".format(
+            PROTOCOL, protocol_spec.name, agent_name
         )
+    )
 
-        output_path = os.path.join(ctx.cwd, item_type_plural)
-        protocol_generator = ProtocolGenerator(specification_path, output_path)
+    if language == "python":
```
Comment (DavidMinarsch) [inline aea/cli/generate.py:124]: declare a constant and use throughout

File: aea/cli/generate.py [hunk 4]
```diff
@@ -119,17 +156,51 @@ def _generate_item(ctx: Context, item_type: str, specification_path: str) -> Non
...
+        if warning_message is not None:
+            click.echo(warning_message)
+
+        # Add the item to the configurations
+        logger.debug(
+            "Registering the {} into {}".format(PROTOCOL, DEFAULT_AEA_CONFIG_FILE)
+        )
+        existing_id_list.add(
+            PublicId(protocol_spec.author, protocol_spec.name, protocol_spec.version)
+        )
```
Comment (DavidMinarsch) [inline aea/cli/generate.py:192]: What's the point of this?

File: aea/cli/generate.py [hunk 5]
```diff
@@ -119,17 +156,51 @@ def _generate_item(ctx: Context, item_type: str, specification_path: str) -> Non
...
+        ctx.agent_loader.dump(
+            ctx.agent_config, open(os.path.join(ctx.cwd, DEFAULT_AEA_CONFIG_FILE), "w")
+        )
```
Comment (DavidMinarsch) [inline aea/cli/generate.py:195]: No point in adding it to the agent. Only Python packages should be added.

File: aea/protocols/generator/base.py [hunk 6]
```diff
@@ -1914,13 +1925,18 @@ def _init_str(self) -> str:
 
         return init_str
 
-    def generate_protobuf_only_mode(self, run_protolint: bool = True) -> None:
+    def generate_protobuf_only_mode(
+        self, language: str = "python", run_protolint: bool = True
```
Comment (DavidMinarsch) [inline aea/protocols/generator/base.py:1929]: Make `python` a constant and reuse throughout.

File: aea/protocols/generator/base.py [hunk 7]
```diff
@@ -1948,10 +1968,18 @@ def generate_protobuf_only_mode(self, run_protolint: bool = True) -> None:
...
 
-        if not is_valid_protobuf_schema:
-            # Remove the generated folder and files
-            shutil.rmtree(output_folder)
-            raise SyntaxError("Error in the protocol buffer schema code:\n" + msg)
+        # Run black formatting
+        try_run_black_formatting(self.path_to_generated_protocol_package)
+
+        # Run isort formatting
+        try_run_isort_formatting(self.path_to_generated_protocol_package)
```
Comment (DavidMinarsch) [inline aea/protocols/generator/base.py:1975]: Why would we run this as part of the protobuf only mode?

File: tests/test_cli/test_generate/test_generate.py [hunk 8]
```diff
@@ -21,7 +21,7 @@
 
 from click import ClickException
 
-from aea.cli.generate import _generate_item
+from aea.cli.generate import protocol
```
Comment (DavidMinarsch) [inline tests/test_cli/test_generate/test_generate.py:24]: That's a problematic method name. Tells where little what it does. I preferred `_generate_item/protocol`

File: aea/cli/generate.py (hunk 1 above)
Comment (5A11) [reply aea/cli/generate.py:56]: Done

File: aea/cli/generate.py (hunk 2 above)
Comment (5A11) [reply aea/cli/generate.py:51]: Done

File: aea/cli/generate.py (hunk 4 above)
Comment (5A11) [reply aea/cli/generate.py:192]: Removed

File: aea/cli/generate.py (hunk 5 above)
Comment (5A11) [reply aea/cli/generate.py:195]: Removed

File: aea/protocols/generator/base.py (hunk 6 above)
Comment (5A11) [reply aea/protocols/generator/base.py:1929]: Done

File: aea/protocols/generator/base.py (hunk 7 above)
Comment (5A11) [reply aea/protocols/generator/base.py:1975]: For the `*_pb2.py` file

Comment (DavidMinarsch) [reply aea/protocols/generator/base.py]: Should only run for the python ones though.

File: aea/cli/generate.py [hunk 9]
```diff
@@ -45,69 +55,118 @@ def generate(
...
+        public_id.name for public_id in existing_protocol_ids_list
+    ]
+    protocol_spec = protocol_generator.protocol_specification
+    protocol_directory_path = os.path.join(ctx.cwd, protocol_plural, protocol_spec.name)
     logger.debug(
         "{} already supported by the agent: {}".format(
-            item_type_plural, existing_item_list
+            protocol_plural, existing_protocol_name_list
         )
     )
```
Comment (DavidMinarsch) [inline aea/cli/generate.py:118]: Remove

**Review by DavidMinarsch — COMMENTED** [review]
LGTM. Please add JS as an option too. Please also extend the docs https://docs.fetch.ai/aea/protocol-generator/ to explain how the new functionality works. And point to it from language agnostic definitions. And link to Go add-on from docs.

File: aea/cli/generate.py (hunk 9 above)
Comment (5A11) [reply aea/cli/generate.py:118]: The whole `logger.debug` block?

Comment (DavidMinarsch) [reply aea/cli/generate.py:118]: Yes, it's factually incorrect in this position.

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**